File: web_blast.pl
#!/usr/bin/perl

# $Id: web_blast.pl,v 1.10 2016/07/13 14:32:50 merezhuk Exp $
#
# ===========================================================================
#
#                            PUBLIC DOMAIN NOTICE
#               National Center for Biotechnology Information
#
# This software/database is a "United States Government Work" under the
# terms of the United States Copyright Act.  It was written as part of
# the author's official duties as a United States Government employee and
# thus cannot be copyrighted.  This software/database is freely available
# to the public for use. The National Library of Medicine and the U.S.
# Government have not placed any restriction on its use or reproduction.
#
# Although all reasonable efforts have been taken to ensure the accuracy
# and reliability of the software and data, the NLM and the U.S.
# Government do not and cannot warrant the performance or results that
# may be obtained by using this software or data. The NLM and the U.S.
# Government disclaim all warranties, express or implied, including
# warranties of performance, merchantability or fitness for any particular
# purpose.
#
# Please cite the author in any work or product based on this material.
#
# ===========================================================================
#
# This code is for example purposes only.
#
# Please refer to https://ncbi.github.io/blast-cloud/dev/api.html
# for a complete list of allowed parameters.
#
# Please do not submit or retrieve more than one request every two seconds.
#
# Results will be kept at NCBI for 24 hours. For best batch performance,
# we recommend that you submit requests after 2000 EST (0100 GMT) and
# retrieve results before 0500 EST (1000 GMT).
#
# ===========================================================================
#
# return codes:
#     0 - success
#     1 - invalid arguments
#     2 - no hits found
#     3 - rid expired
#     4 - search failed
#     5 - unknown error
#
# ===========================================================================

# use strict;
# use warnings;
use URI::Escape;
use LWP::UserAgent;
use HTTP::Request::Common qw(POST);
use JSON qw(decode_json encode_json);
use List::Util qw(max sum);
use File::Basename;

my $ua = LWP::UserAgent->new;

my $argc = $#ARGV + 1;

if ($argc < 3)
    {
    print "usage: web_blast.pl program database query [query]...\n";
    print "where program = megablast, blastn, blastp, rpsblast, blastx, tblastn, tblastx\n\n";
    print "example: web_blast.pl blastp nr protein.fasta\n";
    print "example: web_blast.pl rpsblast cdd protein.fasta\n";
    print "example: web_blast.pl megablast nt dna1.fasta dna2.fasta\n";

    exit 1;
}

my $program = shift;
my $database = shift;

if ($program eq "megablast")
    {
    $program = "blastn&MEGABLAST=on";
    }

if ($program eq "rpsblast")
    {
    $program = "blastp&SERVICE=rpsblast";
    }

# read and encode the queries
my ($basename, $sequence);
my $encoded_query = '';
my $first = 1;

foreach my $query (@ARGV) {
    open(my $fh, '<', $query) or die "Cannot open $query: $!";
    while (<$fh>) {
        $encoded_query .= uri_escape($_);
    }
    close $fh;

    if ($first) {
        # Get the filename without extension
        ($basename) = fileparse($query, qr/\.[^.]*/);

        # Re-open to extract sequence
        open(my $seq_fh, '<', $query) or die "Cannot open $query: $!";
        my @lines = <$seq_fh>;
        close $seq_fh;

        shift @lines if $lines[0] =~ /^>/;
        $sequence = join('', @lines);
        $sequence =~ s/[\r\n]//g;

        $first = 0;
    }
}

# build the request
my $args = "CMD=Put&PROGRAM=$program&DATABASE=$database&QUERY=" . $encoded_query;

# print STDERR "Running query with params:" . $args . "\n";

$ENV{HTTPS_CA_DIR}    = '/etc/ssl/certs';
$ENV{HTTPS_CA_FILE}    = '/etc/ssl/certs/ca-certificates.crt';
$ua->ssl_opts(verify_hostname => 0);

my $req = new HTTP::Request GET => 'https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?' . $args;
# $req->content_type('application/x-www-form-urlencoded');
# $req->content($args);

# get the response
my $response = $ua->request($req);

# print STDERR "Request submission response is : " . $response->content . "\n\n";

# parse out the request id
$response->content =~ /^    RID = (.*$)/m;
my $rid=$1;

# parse out the estimated time to completion
$response->content =~ /^    RTOE = (.*$)/m;
my $rtoe=$1;

print STDERR "Polling response with ID " . $rid . " and estimated response time " . $rtoe . " sec\n";

# wait for search to complete
sleep $rtoe;

# poll for results
while (true)
    {
    sleep 5;

    $req = new HTTP::Request GET => "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_OBJECT=SearchInfo&RID=$rid";
    $response = $ua->request($req);

    # print STDERR "Request status is : " . $response->content . "\n";

    if ($response->content =~ /\s+Status=WAITING/m)
        {
        print STDERR "Searching...\n";
        next;
        }

    if ($response->content =~ /\s+Status=FAILED/m)
        {
        print STDERR "Search $rid failed; please report to blast-help\@ncbi.nlm.nih.gov.\n";
        exit 4;
        }

    if ($response->content =~ /\s+Status=UNKNOWN/m)
        {
        print STDERR "Search $rid expired.\n";
        exit 3;
        }

    if ($response->content =~ /\s+Status=READY/m) 
        {
            # print STDERR "Ready status content: " . $response->content . "\n";
        # if ($response->content =~ /\s+ThereAreHits=yes/m)
            # {
             print STDERR "Search complete, retrieving results...\n";
            last;
            # }
        # else
            # {
            # print STDERR "No hits found.\n";
            # exit 2;
            # }
        }

    # if we get here, something unexpected happened.
    exit 5;
    } # end poll loop

# retrieve and display results
$req = new HTTP::Request GET => "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_TYPE=JSON2_S&RID=$rid";
$response = $ua->request($req);

# print STDERR "Result content: " . $response->decoded_content . "\n";

# Decode the JSON
my $json_text = $response->decoded_content;
my $data = decode_json($json_text);

# Extract the first hit
my $hit = $data->{BlastOutput2}[0]{report}{results}{search}{hits}[0];
my $query_len = $data->{BlastOutput2}[0]{report}{results}{search}{query_len};

# Extract values
my $description     = $hit->{description}[0]{title};
my $scientific_name = $hit->{description}[0]{sciname};
my $accession = $hit->{description}[0]{accession};
my @hsps            = @{ $hit->{hsps} };

my $max_bit_score     = max map { $_->{bit_score} } @hsps;
my $total_bit_score   = sum map { $_->{bit_score} } @hsps;
my $total_score       = sum map { $_->{score} } @hsps;
my $total_identity    = sum map { $_->{identity} } @hsps;

my $query_cover      = sprintf("%.1f%%", 100 * $total_score / $query_len);
my $percent_identity = sprintf("%.1f%%", 100 * $total_identity / $query_len);

my $e_value           = $hsps[0]{evalue};
my $accession_length  = $hit->{len};

# Build the result hash
# my $result = {
#     "description"       => $description,
#     "scientific name"   => $scientific_name,
#     "max score"         => $max_bit_score,
#     "total score"       => $total_bit_score,
#     "query cover"       => $query_cover,
#     "e value"           => $e_value,
#     "percent identity"  => $percent_identity,
#     "accession length"  => $accession_length,
# };

# # Output the result as JSON
# my $json = JSON->new->pretty->allow_nonref;
# print $json->encode($result);

# Prepare fields in order
my @fields = (
    $basename,
    $description,
    $scientific_name,
    $max_bit_score,
    $total_bit_score,
    sprintf("%.1f%%", 100 * $total_score / $query_len),
    $e_value,
    sprintf("%.1f%%", 100 * $total_identity / $query_len),
    $accession_length,
    $accession,
    $sequence
);

# Escape quotes and wrap strings if necessary (simplified CSV formatting)
@fields = map {
    if (/[,"]/) { # if field contains a comma or quote
        s/"/""/g;  # escape quotes by doubling them
        qq("$_")   # wrap in quotes
    } else {
        $_
    }
} @fields;

# Print as a single CSV line
print join(",", @fields), "\n";

exit 0;

File: web_blast.py
#!/usr/bin/env python3
import sys
import os
import re
import time
import json
import requests
from urllib.parse import quote
from functools import reduce

# This script performs a BLAST query using NCBI's web service.
# It replicates the Perl script "web_blast.pl" but in Python 3.

USAGE = """usage: web_blast.py [--output json|csv] program database query [query]...
where program = megablast, blastn, blastp, rpsblast, blastx, tblastn, tblastx

example: web_blast.py blastp nr protein.fasta
example: web_blast.py rpsblast cdd protein.fasta
example: web_blast.py megablast nt dna1.fasta dna2.fasta
"""

BLAST_BASE = "https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi"


def csv_escape(value):
    # Escape quotes and wrap strings if necessary (simplified CSV formatting)
    s = str(value)
    if any(c in s for c in [',', ';', '"']):
        s = s.replace('"', '""')
        return f'"{s}"'
    return s


def read_first_fasta_info(path):
    # Get the filename without extension
    base = os.path.splitext(os.path.basename(path))[0]

    # Read all lines of the FASTA file
    with open(path, 'r') as fh:
        lines = fh.readlines()

    if lines and lines[0].startswith('>'):
        lines = lines[1:]

    seq = ''.join(lines)
    seq = re.sub(r'[\r\n]', '', seq)
    return base, seq


def parse_arguments(argv):
    # Check for optional --output flag
    output_format = "csv"
    args = argv[:]

    if args and args[0] == "--output":
        if len(args) < 2:
            sys.stderr.write(USAGE)
            sys.exit(1)
        output_format = args[1]
        if output_format not in ("csv", "json"):
            sys.stderr.write("Invalid output format. Use 'csv' or 'json'.\n")
            sys.exit(1)
        args = args[2:]

    if len(args) < 3:
        sys.stderr.write(USAGE)
        sys.exit(1)

    program = args[0]
    database = args[1]
    queries = args[2:]
    return output_format, program, database, queries


def main():
    output_format, program, database, queries = parse_arguments(sys.argv[1:])

    if program == "megablast":
        program = "blastn"
        params_extra = {"MEGABLAST": "on"}
    elif program == "rpsblast":
        program = "blastp"
        params_extra = {"SERVICE": "rpsblast"}
    else:
        params_extra = {}

    encoded_query = ''
    first = True
    basename = ""
    sequence = ""

    for q in queries:
        with open(q, 'r') as fh:
            for line in fh:
                encoded_query += quote(line)

        if first:
            basename, sequence = read_first_fasta_info(q)
            first = False

    args = {
        "CMD": "Put",
        "PROGRAM": program,
        "DATABASE": database,
        "QUERY": encoded_query,
    }
    args.update(params_extra)

    headers = {"User-Agent": "web_blast.py (Python requests)"}
    r = requests.get(BLAST_BASE, params=args, headers=headers)

    if not r.ok:
        sys.stderr.write(f"Failed to submit request: HTTP {r.status_code}\n")
        sys.exit(1)

    response_text = r.text

    # parse out the request id
    rid_match = re.search(r'^\s*RID\s*=\s*(\S+)', response_text, re.M)
    rid = rid_match.group(1) if rid_match else None

    # parse out the estimated time to completion
    rtoe_match = re.search(r'^\s*RTOE\s*=\s*(\d+)', response_text, re.M)
    rtoe = int(rtoe_match.group(1)) if rtoe_match else None

    if not rid or rtoe is None:
        sys.stderr.write("Could not parse RID or RTOE\n")
        sys.exit(1)

    print(f"Polling response with ID {rid} and estimated response time {rtoe} sec", file=sys.stderr)
    time.sleep(rtoe / 2)

    while True:
        time.sleep(5)
        poll_args = {"CMD": "Get", "FORMAT_OBJECT": "SearchInfo", "RID": rid}
        r = requests.get(BLAST_BASE, params=poll_args, headers=headers)
        response_text = r.text

        if re.search(r'\s+Status=WAITING', response_text):
            print("Searching...", file=sys.stderr)
            continue
        if re.search(r'\s+Status=FAILED', response_text):
            print(f"Search {rid} failed; please report to blast-help@ncbi.nlm.nih.gov.", file=sys.stderr)
            sys.exit(4)
        if re.search(r'\s+Status=UNKNOWN', response_text):
            print(f"Search {rid} expired.", file=sys.stderr)
            sys.exit(3)
        if re.search(r'\s+Status=READY', response_text):
            print("Search complete, retrieving results...", file=sys.stderr)
            break
        sys.exit(5)

    # retrieve and display results
    result_args = {"CMD": "Get", "FORMAT_TYPE": "JSON2_S", "RID": rid}
    r = requests.get(BLAST_BASE, params=result_args, headers=headers)
    data = r.json()

    hit = data["BlastOutput2"][0]["report"]["results"]["search"]["hits"][0]
    query_len = data["BlastOutput2"][0]["report"]["results"]["search"]["query_len"]

    description = hit["description"][0]["title"]
    scientific_name = hit["description"][0].get("sciname")
    accession = hit["description"][0]["accession"]
    hsps = hit["hsps"]

    max_bit_score = max(h["bit_score"] for h in hsps)
    total_bit_score = sum(h["bit_score"] for h in hsps)
    total_score = sum(h["score"] for h in hsps)
    total_identity = sum(h["identity"] for h in hsps)
    total_align_length = sum(h["align_len"] for h in hsps)

    query_cover = f"{(100 * total_align_length / query_len):.0f}%"
    percent_identity = f"{(100 * total_identity / total_align_length):.2f}%"

    e_value = hsps[0]["evalue"]
    accession_length = hit["len"]

    # Build the result dict
    result = {
        "basename": basename,
        "description": description,
        "scientific_name": scientific_name,
        "max_score": max_bit_score,
        "total_score": total_bit_score,
        "query_cover": query_cover,
        "e_value": e_value,
        "percent_identity": percent_identity,
        "accession_length": accession_length,
        "accession": accession,
        "query_length": query_len,
        "sequence": sequence
    }

    if output_format == "json":
        print(json.dumps(result, indent=2))
    else:
        # CSV
        fields = [
            basename, description, scientific_name,
            max_bit_score, total_bit_score, query_cover,
            e_value, percent_identity, accession_length,
            accession, query_len, sequence
        ]
        fields = [csv_escape(v) for v in fields]
        print(",".join(fields))

    sys.exit(0)


if __name__ == "__main__":
    sys.exit(main())

